kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,375 @@
+package postgres
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteBucket stores the bucket id and the buckets transaction.
+type readWriteBucket struct {
+	// id is used to identify the bucket. If id is null, it refers to the
+	// root bucket.
+	id *int64
+
+	// tx holds the parent transaction.
+	tx *readWriteTx
+
+	table string
+}
+
+// newReadWriteBucket creates a new rw bucket with the passed transaction
+// and bucket id.
+func newReadWriteBucket(tx *readWriteTx, table string, id *int64) *readWriteBucket {
+	return &readWriteBucket{
+		id:    id,
+		tx:    tx,
+		table: table,
+	}
+}
+
+// NestedReadBucket retrieves a nested read bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) walletdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+func parentSelector(id *int64) string {
+	if id == nil {
+		return "parent_id IS NULL"
+	}
+	return fmt.Sprintf("parent_id=%v", *id)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket. This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	cursor := b.ReadWriteCursor()
+
+	k, v := cursor.First()
+	for k != nil {
+		err := cb(k, v)
+		if err != nil {
+			return err
+		}
+
+		k, v = cursor.Next()
+	}
+
+	return nil
+}
+
+// Get returns the value for the given key. Returns nil if the key does
+// not exist in this bucket.
+func (b *readWriteBucket) Get(key []byte) []byte {
+	// Return nil if the key is empty.
+	if len(key) == 0 {
+		return nil
+	}
+
+	var value *[]byte
+	err := b.tx.tx.QueryRowContext(
+		b.tx.db.ctx,
+		"SELECT value FROM "+b.table+" WHERE "+parentSelector(b.id)+" AND key=$1",
+		key,
+	).Scan(&value)
+
+	if err == sql.ErrNoRows {
+		return nil
+	}
+
+	// TODO: In the error case, it is probably not a safe behavior to return
+	// nil. Also applies to etcd.
```

Comment:
@bhandras what's your thought on this?

Reply:
Because adding an error return value at this level would cause a massive refactor, we discussed another possible workaround for this case.

Recap of the offline discussion:
 - In case of an unrecoverable error inside any `Get()` or `Put()` operation (or anything within a TX), we use `panic()`.
 - The outer `kvdb.View()`, `kvdb.Update()` or `kvdb.Batch()` method tries to `recover()` from the panic, logs it, attempts to call `Rollback()` (which will most likely also fail, depending on the error), then attempts to shut down `lnd` as cleanly as possible (logging any message with `Critical()` should be caught by the shutdown logger and do that for us).

The idea is to only do this for etcd and Postgres where the chance of a network failure or similar is way higher than anything `bbolt` and local disk related.

Reply:
Implemented the `recover()` part of this.

For the shutdown logger, do you mean to log a `Critical` message in the postgres driver or let the caller do that? I think if you do it in the postgres driver, the actual error is also bubbled up to the caller. So there will be two error handling processes running in parallel?

Reply:
Decided offline to log a critical message in the recover handler. Test what happens if this actually happens on a running lnd.

Reply:
Added critical logger. Test manually with a delay in the `Get()` function. During the delay, I shut down postgres. Behavior seems to be what we want:

```
2021-07-15 08:19:09.929 [DBG] CHDB: GET DELAY!!!!!
2021-07-15 08:19:19.929 [CRT] CHDB: Caught unhandled error: unexpected EOF
2021-07-15 08:19:19.929 [INF] CHDB: Sending request for shutdown
2021-07-15 08:19:19.929 [INF] LTND: Received shutdown request.
2021-07-15 08:19:19.929 [INF] LTND: Shutting down...
2021-07-15 08:19:19.929 [INF] LTND: Gracefully shutting down.
2021-07-15 08:19:19.929 [ERR] LTND: unable to open graph DB: unexpected EOF
2021-07-15 08:19:19.930 [INF] LTND: Shutdown complete
```